### PR TITLE
Problem: memiavl background snapshot writing don't limit concurrency level 

### PR DESCRIPTION
**Pull request by jacksoom**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/db.go [hunk 1]
```diff
@@ -11,13 +11,15 @@ import (
...
 	"github.com/cosmos/iavl"
 	"github.com/tidwall/wal"
 )
 
 const (
-	DefaultSnapshotInterval = 1000
-	LockFileName            = "LOCK"
+	DefaultSnapshotInterval    = 1000
+	LockFileName               = "LOCK"
+	DefaultSnapshotWriterLimit = 32
```
Comment (yihuang) [inline memiavl/db.go:22]: Shall we use a smaller default value, maybe just 4

File: memiavl/export.go [hunk 2]
```diff
@@ -15,6 +15,8 @@ var ErrorExportDone = errors.New("export is complete")
 // especially since callers may export several IAVL stores in parallel (e.g. the Cosmos SDK).
 const exportBufferSize = 32
 
+const exportSnapshotWorkerNum = 32
```
Comment (yihuang) [inline memiavl/export.go:18]: Can we reuse the default option value?

File: memiavl/db.go (hunk 1 above)
Comment (jacksoom) [reply memiavl/db.go]: > Shall we use a smaller default value, maybe just 4

fixed

File: memiavl/export.go (hunk 2 above)
Comment (jacksoom) [reply memiavl/export.go]: > Can we reuse the default option value?

fixed

File: memiavl/db.go [hunk 3]
```diff
@@ -604,10 +617,12 @@ func (db *DB) Copy() *DB {
...
 	mtree := db.MultiTree.Copy(cacheSize)
+
 	return &DB{
-		MultiTree: *mtree,
-		logger:    db.logger,
-		dir:       db.dir,
+		MultiTree:          *mtree,
+		logger:             db.logger,
+		dir:                db.dir,
+		snapshotWriterPool: db.snapshotWriterPool,
```
Comment (yihuang) [inline memiavl/db.go:625]: we should make the snapshot rewriting goroutines cancellable, and close the pool when closing the db, but that's a separate issue.

**Review by yihuang — APPROVED** [review]

**Comment by yihuang** [comment]
@jacksoom can you run tidy on package `store` and `versiondb`.

**Pull request merged**